src/ImageSharp/Formats/Png/PngDecoderCore.cs

```diff
@@ -391,19 +395,19 @@ private bool TryScaleUpTo8BitArray(ReadOnlySpan<byte> source, int bytesPerScanli
                 return false;
             }
 
-            buffer = this.MemoryAllocator.AllocateManagedByteBuffer(bytesPerScanline * 8 / bits, AllocationOptions.Clean);
-            byte[] result = buffer.Array;
+            buffer = this.memoryAllocator.AllocateManagedByteBuffer(bytesPerScanline * 8 / bits, AllocationOptions.Clean);
+            ref byte sourceRef = ref MemoryMarshal.GetReference(source);
+            ref byte resultRef = ref buffer.Array[0];
             int mask = 0xFF >> (8 - bits);
             int resultOffset = 0;
 
-            for (int i = 0; i < bytesPerScanline - 1; i++)
+            for (int i = 0; i < bytesPerScanline; i++)
             {
-                byte b = source[i];
+                byte b = Unsafe.Add(ref sourceRef, i);
```

Comment:
I wonder if it would not be more efficient to just increment by 1 instead?

Reply:
We avoid index checks using Unsafe